File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& arr) {
        int n = arr.size();
        vector<vector<int>> ans;
        sort(arr.begin() , arr.end());

        for(int i=0; i<n-2 ; i++) {
            if(arr[i] > 0) break;

            if(i > 0 && arr[i] == arr[i-1])  continue;

            if(arr[i] + arr[i+1] + arr[i+2] > 0) break;
            if(arr[i] + arr[n-1] + arr[n-2] < 0) continue;
 
            int j = i+1;
            int k = n-1;

            while(j<k) {
                int sum = arr[i] + arr[j] + arr[k];
                if(sum < 0) j++ ;
                else if (sum > 0) k-- ;

                else {
                    vector<int> temp = {arr[i], arr[j], arr[k]};
                    ans.push_back(temp);

                    j++ ;
                    k-- ;

                    while( j< k && arr[j] == arr[j-1]) j++ ;
                    while( j < k && arr[k] == arr[k+1]) k-- ;
                }
            }   
        }
        return ans;
    }
};

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int n = nums.length;
        Arrays.sort(nums);

        List<List<Integer>> ans = new ArrayList<>();

        for(int i=0; i<n-2 ; i++) {
            if(nums[i] > 0) break;

            if(i>0 && nums[i] == nums[i-1]) continue;

            if(nums[i] + nums[i+1] + nums[i+2] > 0) break;
            if(nums[i] + nums[n-1] + nums[n-2] < 0) continue;

            int j = i+1;
            int k = n-1;

            while(j<k) {
                int sum = nums[i] + nums[j] + nums[k];
                 

                if(sum < 0) j++ ;
                else if (sum > 0) k--;
                else {
                    List<Integer> temp = Arrays.asList(nums[i], nums[j], nums[k]);
                    ans.add(temp);

                    j++ ;
                    k-- ;

                    while(j<k && nums[j] == nums[j-1]) j++ ;
                    while(j<k && nums[k] == nums[k+1]) k-- ;
                }
            }
        }
        return ans;
    }
}